### PR TITLE
WIP: find impl

### PR DESCRIPTION
**Pull request by Sakapoi**

### CODE REVIEW

File: file/Cargo.toml [hunk 1]
```diff
@@ -13,6 +13,8 @@ gettext-rs.workspace = true
 libc.workspace = true
 atty.workspace = true
 regex.workspace = true
+walkdir = "2"
+users = "0.11.0"
```
Comment (SupperZum) [inline file/Cargo.toml:17]: write "0.11" pls, workspace version style

File: file/src/find.rs [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+    Type(char),
+    NoUser,
+    NoGroup,
+    XDev,
+    Prune,
+    Perm(u32),
+}
+
+fn parse_expression(tokens: &mut Vec<&str>) -> Option<Expr> {
+    let mut stack: Vec<Expr> = Vec::new();
```
Comment (SupperZum) [inline file/src/find.rs:33]: write doc pls for all functions

File: file/src/find.rs [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+            if let Ok(metadata) = entry.metadata() {
+                metadata.permissions().mode() & 0o777 == *perm
+            } else {
+                false
+            }
+        }
+    }
+}
+
+fn main() -> ExitCode {
```
Comment (SupperZum) [inline file/src/find.rs:180]: start main from this lines
```
 textdomain(PROJECT_NAME)?;
 bind_textdomain_codeset(PROJECT_NAME, "UTF-8")?;
```

Comment (SupperZum) [inline file/src/find.rs:180]: pls write main signature 
```
fn main() -> Result<(), Box<dyn std::error::Error>>
```

File: file/src/find.rs [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+    let args: Vec<String> = env::args().collect();
```
Comment (SupperZum) [inline file/src/find.rs:181]: use clap for parsing args, pls

File: file/src/find.rs
Comment (SupperZum) [inline file/src/find.rs:1]: add a few testі first pls

**Review by SupperZum — COMMENTED** [review]
In general, please make main() according to the general template. As minimalist as possible.
```rust
fn main() -> Result<(), Box<dyn std::error::Error>> {
    textdomain(PROJECT_NAME)?;
    bind_textdomain_codeset(PROJECT_NAME, "UTF-8")?;
    let mut args = Args::parse();
    args.validate_args()?;
    let mut exit_code = 0;

    if let Err(err) = od(&args) {
        exit_code = 1;
        eprint!("{}", err);
    }

    std::process::exit(exit_code)
}
```

File: file/src/find.rs [hunk 5]
```diff
@@ -0,0 +1,400 @@
...
+    let mut stack: Vec<Expr> = Vec::new();
+
+    while let Some(&token) = tokens.last() {
+        match token {
+            "-name" => {
+                tokens.pop();
+                if let Some(name) = tokens.pop() {
+                    stack.push(Expr::Name(name.to_string()));
+                }
+            }
```
Comment (0x07C0) [inline file/src/find.rs:60]: -name arg doesn't accept the name of a file (or even a part of the name). It accepts a pattern that is a bit more complex to match by: https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_13

File: file/src/find.rs [hunk 6]
```diff
@@ -0,0 +1,400 @@
...
+            "-path" => {
+                tokens.pop();
+                if let Some(path) = tokens.pop() {
+                    stack.push(Expr::Path(path.to_string()));
+                }
+            }
```
Comment (0x07C0) [inline file/src/find.rs:66]: Same as -name

File: file/src/find.rs [hunk 7]
```diff
@@ -0,0 +1,400 @@
...
+                Expr::Name(name) => {
+                    if !file.file_name().to_string_lossy().contains(name) {
+                        c_files.remove(file.path());
+                    }
+                },
+                Expr::Path(path) => {
+                    if !file.path().to_string_lossy().contains(path) {
+                        c_files.remove(file.path());
+                    }
+                },
```
Comment (0x07C0) [inline file/src/find.rs:206]: See comment about -name and -path args. Simply checking if the filename or path contains a substring is not enough.

File: file/src/find.rs [hunk 8]
```diff
@@ -0,0 +1,400 @@
...
+
+#[derive(Debug)]
+enum Expr {
+    And(Vec<Expr>, Vec<Expr>),
+    Or(Vec<Expr>, Vec<Expr>),
+    Not(Vec<Expr>),
+    Name(String),
+    MTime(i64),
+    Path(String),
+    Type(char),
```
Comment (0x07C0) [inline file/src/find.rs:27]: Use a concrete type (enum preferably)

File: file/src/find.rs [hunk 9]
```diff
@@ -0,0 +1,400 @@
...
+
+    if result.is_ok() {
+        for res in result.unwrap() {
+            println!("{}", res.display())
+        }
+        Ok(())
+    }
+    else {
+        Err(result.err().unwrap())
+    }
```
Comment (0x07C0) [inline file/src/find.rs:383]: Just propagate the error here.
```suggestion
        for res in result? {
            println!("{}", res.display())
        }
        Ok(())
```

File: file/src/find.rs [hunk 10]
```diff
@@ -0,0 +1,400 @@
...
+/// # Returns
+///
+/// * A `String` representing the root path extracted from the expressions.
+fn get_root(expr: &[Expr]) -> String {
+    let mut path = String::new();
+    for i in expr {
+        match i {
+            Expr::Path(p) => path = p.to_string(),
+            _ => continue
+        }
```
Comment (0x07C0) [inline file/src/find.rs:346]: Clippy should've caught this. Use `if let` instead or even `iter().find()`.

**Pull request merged**